eth/eth2.go

```diff
@@ -0,0 +1,325 @@
+	}
+	num := parent.Number()
+	header := &types.Header{
+		ParentHash: parent.Hash(),
+		Number:     num.Add(num, common.Big1),
+		Coinbase:   coinbase,
+		GasLimit:   parent.GasLimit(), // Keep the gas limit constant in this prototype
+		Extra:      []byte{},
+		Time:       params.Timestamp,
+	}
+	err = api.eth.Engine().Prepare(bc, header)
+	if err != nil {
+		return nil, err
+	}
+
+	err = api.makeEnv(parent, header)
+	if err != nil {
+		return nil, err
+	}
+	signer := types.NewEIP155Signer(bc.Config().ChainID)
+	txs := types.NewTransactionsByPriceAndNonce(signer, pending)
+
+	var transactions []*types.Transaction
+
+	for {
+		if api.env.gasPool.Gas() < chainParams.TxGas {
+			log.Trace("Not enough gas for further transactions", "have", api.env.gasPool, "want", chainParams.TxGas)
+			break
+		}
+
+		tx := txs.Peek()
+		if tx == nil {
+			fmt.Println("no tx")
+			break
+		}
+
+		from, _ := types.Sender(signer, tx)
+		// XXX replay protection check is missing
+
+		// Execute the transaction
+		api.env.state.Prepare(tx.Hash(), common.Hash{}, api.env.tcount)
+		err := api.commitTransaction(tx, coinbase)
+		switch err {
+		case core.ErrGasLimitReached:
+			// Pop the current out-of-gas transaction without shifting in the next from the account
+			log.Trace("Gas limit exceeded for current block", "sender", from)
+			txs.Pop()
+
+		case core.ErrNonceTooLow:
+			// New head notification data race between the transaction pool and miner, shift
+			log.Trace("Skipping transaction with low nonce", "sender", from, "nonce", tx.Nonce())
+			txs.Shift()
+
+		case core.ErrNonceTooHigh:
+			// Reorg notification data race between the transaction pool and miner, skip account =
+			log.Trace("Skipping account with high nonce", "sender", from, "nonce", tx.Nonce())
+			txs.Pop()
+
+		case nil:
+			// Everything ok, collect the logs and shift in the next transaction from the same account
+			api.env.tcount++
+			txs.Shift()
+			transactions = append(transactions, tx)
+
+		default:
+			// Strange error, discard the transaction and get the next in line (note, the
+			// nonce-too-high clause will prevent us from executing in vain).
+			log.Debug("Transaction failed, account skipped", "hash", tx.Hash(), "err", err)
+			txs.Shift()
+		}
+	}
+
+	block, err := api.eth.Engine().FinalizeAndAssemble(bc, header, api.env.state, transactions, nil /* uncles */, api.env.receipts)
+	if err != nil {
+		return nil, err
+	}
+
+	var logs []*types.Log
+	var receipts = make(types.Receipts, len(api.env.receipts))
+	hash := block.Hash()
+	for i, receipt := range api.env.receipts {
+		// add block location fields
+		receipt.BlockHash = hash
+		receipt.BlockNumber = block.Number()
+		receipt.TransactionIndex = uint(i)
+
+		receipts[i] = new(types.Receipt)
+		*receipts[i] = *receipt
+		// Update the block hash in all logs since it is now available and not when the
+		// receipt/log of individual transactions were created.
+		for _, log := range receipt.Logs {
+			log.BlockHash = hash
```

Comment:
QQ: Looks like the `log.BlockHash` is set to "application-layer" block hash. Do we have any spec for defining the expected behavior?

Reply:
Although the eth2 block hash is not available here. I guess the application layer block hash is the only option here.